File: crates/viben-core/src/channels/discord.rs
//! Discord Channel Client
//!
//! Sends messages via Discord Bot API

use super::types::{DiscordConfig, SendMessageOptions, SendMessageResult, TestChannelResult};
use reqwest::Client;
use serde::{Deserialize, Serialize};

const DISCORD_API_BASE: &str = "https://discord.com/api/v10";

#[derive(Debug, Serialize)]
struct SendMessageRequest {
    content: String,
}

#[derive(Debug, Deserialize)]
struct MessageResponse {
    id: Option<String>,
    message: Option<String>,
}

#[derive(Debug, Deserialize)]
struct UserResponse {
    username: Option<String>,
    discriminator: Option<String>,
    message: Option<String>,
}

/// Send a message to a Discord channel
pub async fn send_discord_message(
    config: &DiscordConfig,
    options: &SendMessageOptions,
) -> SendMessageResult {
    let token = match &config.token {
        Some(t) if !t.is_empty() => t,
        _ => return SendMessageResult::error("Bot token is required"),
    };

    if options.chat_id.is_empty() {
        return SendMessageResult::error("Channel ID is required");
    }

    let client = Client::new();

    let request_body = SendMessageRequest {
        content: options.message.clone(),
    };

    let response = match client
        .post(format!("{}/channels/{}/messages", DISCORD_API_BASE, options.chat_id))
        .header("Authorization", format!("Bot {}", token))
        .header("Content-Type", "application/json")
        .json(&request_body)
        .send()
        .await
    {
        Ok(r) => r,
        Err(e) => return SendMessageResult::error(e.to_string()),
    };

    let status = response.status();
    let data: MessageResponse = match response.json().await {
        Ok(d) => d,
        Err(e) => return SendMessageResult::error(format!("Failed to parse response: {}", e)),
    };

    if !status.is_success() {
        return SendMessageResult::error(
            data.message.unwrap_or_else(|| format!("HTTP {}", status))
        );
    }

    SendMessageResult::success(data.id)
}

/// Test Discord channel configuration by calling /users/@me
pub async fn test_discord_channel(config: &DiscordConfig) -> TestChannelResult {
    let token = match &config.token {
        Some(t) if !t.is_empty() => t,
        _ => return TestChannelResult::error("Bot token is required"),
    };

    let client = Client::new();

    let response = match client
        .get(format!("{}/users/@me", DISCORD_API_BASE))
        .header("Authorization", format!("Bot {}", token))
        .send()
        .await
    {
        Ok(r) => r,
        Err(e) => return TestChannelResult::error(e.to_string()),
    };

    let status = response.status();
    let data: UserResponse = match response.json().await {
        Ok(d) => d,
        Err(e) => return TestChannelResult::error(format!("Failed to parse response: {}", e)),
    };

    if !status.is_success() {
        return TestChannelResult::error(
            data.message.unwrap_or_else(|| format!("HTTP {}", status))
        );
    }

    TestChannelResult::success(format!(
        "Bot: {}#{}",
        data.username.unwrap_or_else(|| "unknown".to_string()),
        data.discriminator.unwrap_or_else(|| "0000".to_string())
    ))
}

File: packages/core/src/channels/discord.ts
/**
 * Discord Channel Client
 *
 * Sends messages via Discord Bot API
 */

import type {
  DiscordChannelConfig,
  SendMessageOptions,
  SendMessageResult,
  TestChannelResult,
} from "./types";
import { fetchWithProxy } from "./http-client";

const DISCORD_API_BASE = "https://discord.com/api/v10";

/**
 * Send a message to a Discord channel
 */
export async function sendDiscordMessage(
  config: DiscordChannelConfig,
  options: SendMessageOptions
): Promise<SendMessageResult> {
  if (!config.token) {
    return { success: false, error: "Bot token is required" };
  }

  if (!options.chatId) {
    return { success: false, error: "Channel ID is required" };
  }

  try {
    const response = await fetchWithProxy(
      `${DISCORD_API_BASE}/channels/${options.chatId}/messages`,
      {
        method: "POST",
        headers: {
          Authorization: `Bot ${config.token}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          content: options.message,
        }),
      }
    );

    const data = (await response.json()) as {
      id?: string;
      message?: string;
    };

    if (!response.ok) {
      return {
        success: false,
        error: data.message || `HTTP ${response.status}`,
      };
    }

    return {
      success: true,
      messageId: data.id,
    };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Test Discord channel configuration by calling /users/@me
 */
export async function testDiscordChannel(
  config: DiscordChannelConfig
): Promise<TestChannelResult> {
  if (!config.token) {
    return { success: false, error: "Bot token is required" };
  }

  try {
    const response = await fetchWithProxy(`${DISCORD_API_BASE}/users/@me`, {
      method: "GET",
      headers: {
        Authorization: `Bot ${config.token}`,
      },
    });

    const data = (await response.json()) as {
      username?: string;
      discriminator?: string;
      message?: string;
    };

    if (!response.ok) {
      return {
        success: false,
        error: data.message || `HTTP ${response.status}`,
      };
    }

    return {
      success: true,
      details: `Bot: ${data.username}#${data.discriminator}`,
    };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}
